### ISSUE TITLE
Does the concept of a compiler fence make any sense?

### ISSUE THREAD

**Issue by RalfJung** [rust-lang/unsafe-code-guidelines#347]
Atomic fences are specified in terms of things that happen when certain *atomic* reads occur:
- on thread A: first a fence, then some (atomic) store
- on thread B: first an (atomic) load, then a fence

In that particular situation, if the load reads-from the store, then the fences kick in and have an effect. That is the only effect they have, I think.

So, if your program contains no atomic accesses, but some atomic fences, those fences do nothing.
We also think that an atomic fence has at least all the effects of a compiler fence, i.e., a compiler fence is strictly weaker than an atomic fence. But that means a compiler fence has no effect on programs without atomic accesses -- which is just wrong, that's not how they are supposed to behave.

So what *is* the operational spec of a compiler fence? I have no idea.
See [the discussion here](https://github.com/rust-lang/unsafe-code-guidelines/issues/321#issuecomment-1175417827) for some more details. Let's go on discussing here.

**Comment by RalfJung** [comment]
So you are going in the direction of what was discussed [around here](https://github.com/rust-lang/unsafe-code-guidelines/issues/321#issuecomment-1087976856).

> These guarantees would probably have to be specified in the same way that FFI or inline assembly is specified - as a relation between the abstract machine state and the underlying lower level state of the machine.

That would block a lot more optimizations that you would want when writing concurrent software with atomics. In particular we want there to be a meaningful difference between atomic fences with different orderings, which I do not think is the case under your proposal. So I don't think atomic fences should have any effect like that.

So maybe the statement that an atomic fence is strictly stronger than a compiler fence is wrong?

> I think this is evidenced by the fact that we would expect atomic access from Rust code to be able to synchronize with atomic accesses from another language (or indeed assembly) in the same program. 

I don't agree. Atomics are a feature inside the language and its memory model, so one should only expect them to synchronize inside that memory model. So whatever the other code does must be cast in terms of that memory model.

C++ atomics are *not* some abstract way of talking about what the hardware does, they are their completely own thing that is then *implemented* in terms of what the hardware does. There are sometimes even multiple different mutually incompatible schemes to implement the same model on a given hardware!

@DemiMarie (in the old thread)
> Are there semantics for interrupts at all? That is a prerequisite for having semantics for compiler fences.

I guess part of the question here is whether interrupts exist as a language feature (that we'd have to add to the AM) and compiler fences interact with that, or whether we consider them to be *outside* the AM and compiler-fence to be a way to somehow coordinate with out-of-AM activities.

Atomic operations are *not* an out-of-AM mechanism. This is one of the many key differences between atomic accesses and volatile accesses. So it would be very strange to me if atomic fences were ouf-of-AM mechanisms, and indeed that would put a lot of extra cost on concurrent algorithms that are perfectly happy staying entirely inside the AM.

**Comment by DemiMarie** [comment]
> > I think this is evidenced by the fact that we would expect atomic access from Rust code to be able to synchronize with atomic accesses from another language (or indeed assembly) in the same program.
> 
> I don't agree. Atomics are a feature inside the language and its memory model, so one should only expect them to synchronize inside that memory model. So whatever the other code does must be cast in terms of that memory model.

Cross-FFI atomics need to work.  The question is how to specify atomics in such a way that they do.

**Comment by Diggsey** [comment]
> In particular we want there to be a meaningful difference between atomic fences with different orderings, which I do not think is the case under your proposal. So I don't think atomic fences should have any effect like that.

That seems like a bit of a leap? I assume you are referring to:
>  There are sometimes even multiple different mutually incompatible schemes to implement the same model on a given hardware!

I would expect atomics to be like part of the ABI - ie. the compiler is not expected to make *any* kind of atomic accesses from FFI code work correctly, but as long as the FFI code follows the same rules (for that ordering) as the Rust compiler does, then they would be expected to synchronize. I don't see why this would make orderings redundant.

**Comment by Diggsey** [comment]
> I don't agree. Atomics are a feature inside the language and its memory model, so one should only expect them to synchronize inside that memory model. So whatever the other code does must be cast in terms of that memory model.

It's my understanding that atomics are the expected way to do IPC via shared memory. I don't see how to reconsile your statement with that.

**Comment by RalfJung** [comment]
> It's my understanding that atomics are the expected way to do IPC via shared memory. I don't see how to reconcile your statement with that. (This is certainly a use-case supported by C++'s atomics)

For shared memory *with other code that follows the same memory model*.  Basically, with other instances of (roughly) the same AM.

**Comment by Lokathor** [comment]
I'm unclear: are you saying that Rust atomics can sync with C++ atomics, or not?

**Comment by RalfJung** [comment]
> That seems like a bit of a leap? I assume you are referring to:

No that's not what I am referring to. I'm saying `atomic::fence` has this ordering parameter and we agree it is relevant for the semantics, right? Because under your spec I don't see how it is.

> I'm unclear: are you saying that Rust atomics can sync with C++ atomics, or not?

We are using the C++ memory model so that works fine.

But syncing e.g. with assembly code only makes sense if whatever that assembly code does can be expressed in terms of this memory model. You can't expect an atomic fence to have any effect other than what it says in the memory model.

**Comment by RalfJung** [comment]
> I guess part of the question here is whether interrupts exist as a language feature (that we'd have to add to the AM) and compiler fences interact with that, or whether we consider them to be outside the AM and compiler-fence to be a way to somehow coordinate with out-of-AM activities.

To be clear, I think saying that compiler_fence is an out-of-AM syncing mechanism is not a bad idea. However, that would make compiler_fence incomparable in strength with atomic fences (and having `atomic::compiler_fence` just makes no sense).

That is actually consistent with what @comex wrote
> But if we do want to allow reordering, I'd say that compiler_fence should only have a specified effect in conjunction with volatile accesses, similar to how atomic fences only have an effect in conjunction with atomic accesses.

Volatile accesses and atomic accesses are also incomparable in strength.

However, I just realized compiler_fence has an `Ordering`, and I have no idea what that is supposed to mean then...

**Comment by DemiMarie** [comment]
> > I'm unclear: are you saying that Rust atomics can sync with C++ atomics, or not?
> 
> We are using the C++ memory model so that works fine.

I think we should seriously consider looking at the Linux kernel memory model as well.  This has the advantage that it is _known_ to work for low-level programming at scale, whereas I am not aware of the C++ memory model being used for such code.  Rust will need to integrate with the Linux kernel memory model to be used as part of the Linux kernel anyway.

**Comment by Diggsey** [comment]
> I'm saying atomic::fence has this ordering parameter and we agree it is relevant for the semantics, right? Because under your spec I don't see how it is.

Oh I see - because the memory ordering is part of the C++/Rust memory model.

I guess another option would be to add the concept of "interruption" to the abstract machine, and then define compiler fences in terms of that. I think it's possible to define everything about "interruption" except for what triggers it - that's the only part that's "outside" what we can define.

**Comment by m-ou-se** [comment]
> But that means a compiler fence has no effect on programs without atomic accesses -- which is just wrong, that's not how they are supposed to behave.

Why is that wrong?

Single-threaded / signal / compiler fences are relevant for signal handlers, interrupts, and things like Linux' `SYS_membarrier`.

**Comment by Lokathor** [comment]
What's wrong is the "that means a compiler fence has no effect..." part, because they're supposed to have *some sort of effect*.

We all seem to agree they should do *something*.

**Comment by m-ou-se** [comment]
They do have an effect, but not in a program without any atomic operations.

**Comment by Lokathor** [comment]
Ah, then I was incorrect and we seem to disagree.

Many people (myself included) have been under the impression that a compiler_fence has an effect even in a program with no atomic operations (particularly, even for targets where there are no atomics). If that's not the case we *really* need to fix the docs.

**Comment by m-ou-se** [comment]
The C++ standard defines `atomic_signal_fence` (aka our `atomic::compiler_fence`) as:

> Equivalent to atomic_­thread_­fence(order)¹, except that the resulting ordering constraints are established only between a thread and a signal handler executed in the same thread[.](https://eel.is/c++draft/full#atomics.fences-6.sentence-1)

(¹ Aka our `atomic::fence()`.)

And adds:

> [Note [1](https://eel.is/c++draft/full#atomics.fences-note-1): atomic_­signal_­fence can be used to specify the order in which actions performed by the thread become visible to the signal handler[.](https://eel.is/c++draft/full#atomics.fences-7.sentence-1)
Compiler optimizations and reorderings of loads and stores are inhibited in the same way as with atomic_­thread_­fence, but the hardware fence instructions that atomic_­thread_­fence would have inserted are not emitted[.](https://eel.is/c++draft/full#atomics.fences-7.sentence-2)
— end note]

Other types of interrupts or things like SYS_membarrier aren't part of the C++ standard, but are also use cases for a single-threaded/compiler/signal fence.

A more practical/wider applicable way of defining `atomic::compiler_fence` would be to say it is identical to `atomic::fence`, except it doesn't work across hardware threads/cores. That means nothing in the abstract machine of course. So it's just a regular atomic fence, but with an additional assumption/requirement from outside the abstract machine. If the abstract machine defines signal handlers or interrupts, it could possibly mention something about those situations specifically.

**Comment by Lokathor** [comment]
Well, I've never programmed C++, and so honestly I only vaguely follow what that's saying.

I'm unclear on what the implications are for targets without atomics, where we might assume that there's only one core and possibly an interrupt handler.

**Comment by m-ou-se** [comment]
When you say "without atomics", do you mean without atomic read-modify-write/cas operations, or even without `load` and `store`?

If a platform has _no_ atomic load or store for any size atomic.. then there's no way to communicate between a signal handler/interrupt and the main code at all through shared memory. Then there's indeed no use for a fence.

**Comment by Lokathor** [comment]
I mean for example ARMv4T, which can uninterruptedly read-write "swap" with `swp` (u32) and `swpb` (u8), but there's no atomic-compare-and-swap, and no atomic-read-modify-write.

And it does have interrupts.

But you're telling me there's no official/legal way for the interrupt to communicate anything to the main program? That seems... unfortunate.

**Comment by Lokathor** [comment]
Well, there's normal load and store instructions, but not atomic anything instructions.

**Comment by m-ou-se** [comment]
> Well, there's normal load and store instructions, but not atomic anything instructions.

That's the same. All load and store instructions on nearly all architectures are also atomic.

**Comment by Lokathor** [comment]
Oh, well then sure, there's atomic loading and storing.

**Comment by chorman0773** [comment]
BTW, I interpret `core::sync::atomic::compiler_fence` as being Rust's version of `atomic_signal_fence` - that is, a fence that can only synchronize with atomic ops and fences on the same thread of execution as the calling code (including from a signal handler on that thread).

Reading it this way makes quite a bit of sense to me at least - it just modifies how the fence-atomic, atomic-fence, and fence-fence synchronization rules function.

**Comment by Lokathor** [comment]
**Query:** How does one enforce ordering between volatile and non-volatile ops.

Previously, I've *also* seen people assume that `compiler_fence` can do this.

But, again, apparently `compiler_fence` does nothing at all?

**Comment by chorman0773** [comment]
> Query: How does one enforce ordering between volatile and non-volatile ops.

You can't*. 
If the volatile op was also atomic, though, then you could get the same synchronization behaviour in a signal/interrupt handler as a full `fence` would.

And yes, rust needs volatile atomic operations.

**Comment by repnop** [comment]
> How does one enforce ordering between volatile and non-volatile ops.

this is what I've been using `compiler_fence` for, I need _some_ way to prevent the compiler from reordering non-volatile and volatile operations to write driver code, and the docs for `compiler_fence` don't mention anything about atomics until the example, so I was under the impression the reason for it being in `core::sync::atomic` was because it uses the `Ordering` enum from there. looking at the source, it indeed calls atomic fence intrinsics
```rust
#[inline]
#[stable(feature = "compiler_fences", since = "1.21.0")]
#[rustc_diagnostic_item = "compiler_fence"]
pub fn compiler_fence(order: Ordering) {
    // SAFETY: using an atomic fence is safe.
    unsafe {
        match order {
            Acquire => intrinsics::atomic_singlethreadfence_acq(),
            Release => intrinsics::atomic_singlethreadfence_rel(),
            AcqRel => intrinsics::atomic_singlethreadfence_acqrel(),
            SeqCst => intrinsics::atomic_singlethreadfence(),
            Relaxed => panic!("there is no such thing as a relaxed compiler fence"),
        }
    }
}
```
which makes me think that the docs really need improved to mention it only applies to atomic accesses if that's the only guarantee that it makes, and it doesn't help that it links to the Linux memory barrier docs which discuss _much_ more than just atomic orderings. IMO there should be some other way to prevent non-volatile & volatile reordering as its necessary to be able to guarantee that for things like driver code. The Linux kernel has functions specifically for accessing MMIO with barriers to prevent reordering in code, but do we make any such guarantee about `asm!` preventing function calls from being reordered inside the caller?

**Comment by RalfJung** [comment]
> How does one enforce ordering between volatile and non-volatile ops.

Indeed I thought this was one of the motivations for compiler-fence: so that you can do a bunch of regular writes, then a fence, and then a single volatile write to "publish" something via MMIO to some other DMA participant.

So I think we basically have 3 choices here:
1. Provide another kind of fence for that. Then we'll still have to figure out how to spec that fence though, and it'll probably have to be [something like this](https://github.com/rust-lang/unsafe-code-guidelines/issues/321#issuecomment-1087976856). That might just be an empty `asm!` block (with read-only memory clobber, if that's a thing), since it looks like that's what that spec boils down to.
2. Allow using either a compiler-fence or a full atomic fence for that. This however means using atomic fences for atomics imposes volatile-related restrictions that that code does not need.
3. Allow using *only* compiler-fence for that, which means it is no longer always correct to replace a compiler-fence by an atomic fence.

Looks like C/C++ use (1), but don't provide an official way to write that other fence? I am opposed to (2) since impacting purely atomic code like that seems bad.

**Comment by Lokathor** [comment]
read-only asm blocks are indeed a thing.

**Comment by comex** [comment]
> Indeed I thought this was one of the motivations for compiler-fence: so that you can do a bunch of regular writes, then a fence, and then a single volatile write to "publish" something via MMIO to some other DMA participant.

> So I think we basically have 3 choices here:
> 
> 1. Provide another kind of fence for that.

Hold on.  For the "signal handler" use case, the goal is still to provide ordering between non-volatile accesses and a volatile access.  So it's not so different from the MMIO use case.

[This answer linked earlier](https://stackoverflow.com/questions/18449291/when-is-a-compiler-only-memory-barrier-such-as-stdatomic-signal-fence-useful/18454971#18454971) suggests that a compiler fence could be used "by itself" to synchronize access to variables, without mentioning volatile, but if the intent is really that none of the variables in the example are volatile, it's wrong.  You can't use a non-atomic non-volatile variable to synchronize between a thread and its signal handler, even with a compiler fence, for the same reasons you can't use one to synchronize between two threads even with an atomic fence.  The problematic optimizations are the same in both cases.

**Comment by workingjubilee** [comment]
> > Well, there's normal load and store instructions, but not atomic anything instructions.
> 
> That's the same. All load and store instructions on nearly all architectures are also atomic.
> 
> > Aren't these loads and stores atomic, if they are "uninterruptable" and there is no multi-core version of this hardware?
> 
> Indeed.

Reaffirming but also adding a note:
- All **aligned** loads and stores on *functionally* all architectures are also atomic.
- Some architectures allow unaligned load/store and make them atomic.
- For some architectures, the execution of an unaligned load/store allows tearing.

**Comment by DemiMarie** [comment]
People _will_ use write-barriers for things other than MMIO.  Xen’s libvchan uses them for communication between virtual machines, for example.  I can also state that driver writers will use ordinary (non-volatile, non-atomic) loads and stores plus memory barriers, instead of using volatile and/or atomic operations for every single access to shared memory.

**Comment by RalfJung** [comment]
Driver writers are not using the C++ memory model (assuming you are talking about Linux) so that's an apples-to-oranges comparison.

But anyway MMIO is just an example for "things outside the AM" -- "things you'd need `volatile` for".

**Comment by repnop** [comment]
empty `asm!("");`s definitely do not function as a complete fence, optimizations are allowed to use the lack of clobbers as a sign that it doesn't touch the AM state. it does heavily pessimise optimizations from some things I've thrown it at, but for example, aliasing analysis with `&mut` can cause reordering of code because it knows there can't be any aliasing pointers.

**Comment by workingjubilee** [comment]
It is my opinion that relying on any properties of `volatile` beyond the "cache-defeating" property, the property for which it was initially introduced 30 years ago in C89, is unwise. The cache-defeating property has implications for ordering but that doesn't really constitute the same thing on its own. Thus @chorman0773's opinion that we may need `atomic_volatile` is plausible.

At the moment, LLVM's LangRef only defines two fences, `fence` (the atomic kind) and `llvm.arithmetic.fence`. I do not believe LLVM has a model of such sync that is not implicitly about atomic orderings, as a result.

As our compiler fence is `atomic_signal_fence`, it appears, which syncs normal and atomic accesses, combined with the cache-defeating property, I believe LLVM must nonetheless guarantee ordering for certain combinations of volatile and atomic accesses using `compiler_fence`. I also believe certain sequences of volatile reads and writes are naturally forced to sequence, but I say that with low confidence.

**Comment by Lokathor** [comment]
We can confidently say, for what it's worth, that all volatile ops are definitely kept in order relative to other volatile ops. They're implicitly SeqCst relative to other volatile ops, one might say.

And as you say Jubilee, normal data dependencies apply between volatile and standard accesses still apply: If you standard write data that came from a volatile read, the standard write cannot happen before the volatile read simply because it doesn't exist before the read.

**Comment by chorman0773** [comment]
FWIW, XIR has a `fence` and `sequence` instruction, comparable to `atomic_thread_fence` and `atomic_signal_fence` respectively, but that allows any XIR access-class (including merely `volatile`). The current semantics of a `fence volatile` or `sequence volatile` are not entirely worked out, but at a minimum, it will be treated as a side effect and forbid other volatile operations from being reordered relative to it. I think it could easily be defined to forbid the reordering of non-volatile accesses arround them as well.

**Comment by comex** [comment]
> empty `asm!("");`s definitely do not function as a complete fence, optimizations are allowed to use the lack of clobbers as a sign that it doesn't touch the AM state. it does heavily pessimise optimizations from some things I've thrown it at, but for example, aliasing analysis with `&mut` can cause reordering of code because it knows there can't be any aliasing pointers.

The issue isn't the lack of clobbers exactly.  The "new" `asm!` has the `memory` clobber (in the syntax of GCC or the old `asm!`) on by default, unless you pass `nomem` to disable it.

But indeed, that is a caveat I wasn't aware of.  In this test case:
```rust
use std::arch::asm;
pub unsafe fn foo(x: &mut i32, y: *mut *mut i32) {
    *x = 1;
    asm!("");
    y.write_volatile(x);
    // ^-- Imagine that is triggering an MMIO device
    // to start reading from x.
    *x = 2;
}
```
the `*x = 1;` is optimized out.  The `asm!("")` is not enough to establish ordering because that asm block is not allowed to read from `x`.

The optimization can be prevented using `std::sync::atomic::compiler_fence`.  It would also be prevented if volatile accesses were reimplemented using asm themselves.

**Comment by Lokathor** [comment]
Other than the "we would have to write some new code for every arch" problem I don't see anything particularly wrong with implementing volatile with asm blocks.

**Comment by comex** [comment]
> > If the compiler_fence was replaced by an atomic fence, it would definitely be wrong -- we agree on that, right? The same optimizations that break for the atomic case, can also still be done with a compiler_fence, so that version is just as wrong.

I don't know if we disagree on anything.  Even supposing for argument's sake that no optimizations occur involving moving anything across the barrier, and no interprocedural optimizations occur…

Assuming `is_shared_data_initialized` is a `bool`, even if we assume that `bool` is 1 byte on the platform (it's not always) and thus the write to it cannot tear, the generated assembly could write an invalid value to that byte (not 0 or 1) before writing the final value; at that point, when the signal handler reads the value, it may misbehave due to the compiler assuming the value is 0 or 1.

...I did miss the fact that `atomic_signal_fence` is defined to only affect atomics.  It doesn't even affect `volatile sig_atomic_t`, which, as @chorman0773 pointed out, is the *only* case where the C++ spec blesses using `volatile` to communicate between a thread and its signal handler.

But [some](https://github.com/nrf-rs/nrf-hal/blob/663008c033ad67263e4ac0c561d5d1fac28d7170/nrf-hal-common/src/spis.rs#L158) people [definitely](https://github.com/stm32-rs/stm32l4xx-hal/blob/945c6a436ce9d1bacc10e237569ede28ee19361f/src/dma.rs#L775) are using `compiler_fence` for DMA, or in other words to provide a barrier between non-volatile accesses and volatile-but-not-atomic accesses.

Implementation-wise, today, `compiler_fence` prevents undesired optimizations in that case while `asm!("");` may not, per my last comment.  *However,* it's implemented using LLVM's `fence` instruction, which is documented by LangRef as only working with atomics.

**Comment by Lokathor** [comment]
why would there ever be code emitted that temporarily writes and invalid value to a `bool`, and if that can happen for one type of value why can't that randomly also happen for any other type of value? Should I be concerned that my `&mut bool` values might suddenly explode some day?

**Comment by chorman0773** [comment]
Something could use the same storage the `bool` occupies while it knows its
not being used, and it just remembers the value to restore for the next
use because compiler.

**Comment by RalfJung** [comment]
> aliasing analysis with &mut can cause reordering of code because it knows there can't be any aliasing pointers.

That is true with all fences, none of them allows you to violate the aliasing assumptions.
If you are using volatile because that memory is shared with some other party (e.g. via MMIO), then you must also inform the type system of that sharing, by using shared references or raw pointers.

> And as you say Jubilee, normal data dependencies apply between volatile and standard accesses still apply: If you standard write data that came from a volatile read, the standard write cannot happen before the volatile read simply because it doesn't exist before the read.

The compiler might under some circumstances be able to speculate the write. Then such ordering can still be violated.

Reasoning with data dependencies is pretty much impossible in a programming language. Assembly langues can do it but that's a different situation. (This is why the 'consume' memory ordering is just inherently broken.)

> The optimization can be prevented using std::sync::atomic::compiler_fence.

That is coincidence and not guaranteed by any spec. Mutable references are unique and the memory they point to must not be written to or read from through any other pointer (not derived from them). This includes MMIO and DMA accesses of other devices.

**Comment by RalfJung** [comment]
> But [some](https://github.com/nrf-rs/nrf-hal/blob/663008c033ad67263e4ac0c561d5d1fac28d7170/nrf-hal-common/src/spis.rs#L158) people [definitely](https://github.com/stm32-rs/stm32l4xx-hal/blob/945c6a436ce9d1bacc10e237569ede28ee19361f/src/dma.rs#L775) are using compiler_fence for DMA, or in other words to provide a barrier between non-volatile accesses and volatile-but-not-atomic accesses.

Yeah my feeling is that we probably want to say that compiler_fence can be used for this purpose, and achieve that by adding an empty `asm!` block (with `memory` clobber) in its implementation so that LLVM also considers this a barrier.

**Comment by comex** [comment]
> That is coincidence and not guaranteed by any spec. Mutable references are unique and the memory they point to must not be written to or read from through any other pointer (not derived from them). This includes MMIO and DMA accesses of other devices.

Oh, true, my example is wrong.  Here's a better version - again, the goal is to show that `asm!("")` does not always act as a fence between non-volatile and volatile accesses:

```rust
use std::arch::asm;
pub unsafe fn foo(y: *mut *mut i32) {
    let mut val: i32 = 0;
    let x: *mut i32 = &mut val;

    *x = 42; 
    asm!("");
    y.write_volatile(x);
    
    // Wait for DMA to be done, to ensure that `val`
    // doesn't go out of scope first.  This is actually
    // unnecessary, since the system would be entitled
    // to read memory before allowing the volatile
    // access to complete, but this just proves that
    // that isn't the issue here.
    while !y.read_volatile().is_null() {}
}
```

The write of `42` is elided, because from the compiler's perspective there is no way for code to read from it without either UB due to a data race or UB due to reading deallocated memory – given the lack of a memory barrier between where its address is first exposed and where it goes out of scope.  If you replace the `asm!("");` with a proper `compiler_fence(Release)`, then the write is no longer elided.

The reason that `asm!("")` *usually* acts like a compiler fence is that the compiler must assume that *the asm block* might read any memory it could potentially have valid provenance to read.  So it ensures that any prior writes to Abstract Machine memory are actually reflected in "concrete machine" memory as of the asm block.  And those values are still there at the time of the immediately following volatile write.

But in this case, provenance to `x` escapes through the volatile write itself, so the asm block would not be entitled to read it, so the compiler does not synchronize it .

**Comment by comex** [comment]
Oh, and just to reiterate, I understand that the `fence` LLVM IR instruction that `compiler_fence` produces, while it currently provides a barrier, is not guaranteed to do so, since the LangRef doesn't specify it as having *any* effect on volatile non-atomic accesses.  This does make me partial to the idea of re-implementing volatile accesses on top of asm, because that just removes the whole issue.  As an alternative, it seems plausible to me that LLVM could be convinced to change the definition of `fence` to also provide assembly-level ordering between non-volatile accesses and volatile accesses. /shrug

**Comment by Lokathor** [comment]
LLVM Volatile Docs:
> The allowed side-effects for volatile accesses are limited. If a non-volatile store to a given address would be legal, a volatile operation may modify the memory at that address. A volatile operation may not modify any other memory accessible by the module being compiled. A volatile operation may not call any code in the current module.

So if I read that right, a volatile op on its own can't trigger DMA to modify "any other memory accessible by the module being compiled". So you can't DMA into standard memory using volatile alone (you'd have to actually break out the inline assembly).

Which doesn't solve the signal/interrupt handler case, but it reduces the DMA case.

**Comment by comex** [comment]
And yet plenty of code in C and Rust *expects* to be able to do this, maybe not with volatile alone, but at least with volatile plus some sort of barrier...

**Comment by Lokathor** [comment]
"tell llvm they need to change their rules" seems like a taller task than "just use an assembly block already"

**Comment by thalesfragoso** [comment]
> So if I read that right, a volatile op on its own can't trigger DMA to modify "any other memory accessible by the module being compiled". So you can't DMA into standard memory using volatile alone (you'd have to actually break out the inline assembly).

The volatile operation in the DMA case is not modifying any other memory, it's merely escaping the pointer to another "thread" (the DMA), and that thread is the one doing the modifications. That seems okay for me.

However, that still leaves the problem on how to achieve proper synchronization with the DMA "thread" without using atomics (assuming compiler_fence only works with atomics).

**Comment by DemiMarie** [comment]
> Reasoning with data dependencies is pretty much impossible in a programming language. Assembly langues can do it but that's a different situation. (This is why the 'consume' memory ordering is just inherently broken.)

The problem is that current hardware requires such reasoning in order to get maximum performance, unless I have missed something.

**Comment by RalfJung** [comment]
>  If you replace the asm!(""); with a proper compiler_fence(Release), then the write is no longer elided.

Let's imagine this were a full `atomic` fence, and the volatile accesses were relaxed atomic accesses. Now the question is, would LLVM be allowed to elide the write?

I guess not, because another thread could see the value stored in `y`, read `x`, read `*x`, and only then set `*y` to null which will deallocate `x`. In fact this would apply even without the fence.

If that is correct then I think LLVM is wrong to elide the write here, *even if there is no fence of any kind*.

**Comment by Amanieu** [comment]
This code is incorrect, you need a second fence at the end to re-acquire `x` before it goes out of scope. Once you add the second fence the `42` is no longer elided.
```rust
use std::arch::asm;
pub unsafe fn foo(y: *mut *mut i32) {
    let mut val: i32 = 0;
    let x: *mut i32 = &mut val;

    *x = 42; 
    asm!(""); // compiler_fence(Release)
    y.write_volatile(x);
    
    // Wait for DMA to be done, to ensure that `val`
    // doesn't go out of scope first.  This is actually
    // unnecessary, since the system would be entitled
    // to read memory before allowing the volatile
    // access to complete, but this just proves that
    // that isn't the issue here.
    while !y.read_volatile().is_null() {}

    asm!(""); // compiler_fence(Acquire)
}
```

**Comment by RalfJung** [comment]
Ah, things make sense then. :)

**Comment by comex** [comment]
Bah, you're right.  Try 3; again the store is elided:
```rust
use std::arch::asm;
pub unsafe fn foo(y: *mut *mut i32) {
    let mut val: i32 = 0;
    let x: *mut i32 = &mut val;

    *x = 42; 
    asm!(""); // compiler_fence(Release)
    y.write_volatile(x);
    loop {}
}
```

**Comment by RalfJung** [comment]
Might be some remnant of "there are no infinite silent loops" reasoning? Is it still elided when you put some kind of side-effect into the loop?

**Comment by Lokathor** [comment]
How does the compiler know that `asm!("")` is acquire or release like the comments suggest?

**Comment by RalfJung** [comment]
`asm!` is support to be an overapproximation of all these options. @Amanieu said it would even be legal to patch in some other code at runtime as long as it still respects the same clobbers.

However, in this case the state being clobbered is "synchronization with other threads". This state is absolutely real on the Abstract Machine but compilers rarely talk about it. So... does `asm!` clobber that state by default, and if not, then how can I annotate it to make that state clobbered?

**Comment by Amanieu** [comment]
Another example which strongly suggests an issue with LLVM:
```rust
use std::sync::atomic::{*, Ordering::*};

// If this function is inlined then 42 is written,
// otherwise it is elided.
#[inline(never)]
fn bar() {
    compiler_fence(Release);
}

pub unsafe fn foo(y: *mut *mut i32) {
    let mut val: i32 = 0;
    let x: *mut i32 = &mut val;

    *x = 42;
    bar();
    y.write_volatile(x);
    loop {}
}
```

Anyways, getting back on topic, the way I view `compiler_fence` is that it works exactly like `fence` except that it only synchronizes with other threads on the same CPU. In practice "same CPU" include signal handlers or interrupt handlers which act as another thread from the AM's point of view (which also suspends the original thread).

The way you actually use `compiler_fence` in practice is with `Relaxed` atomics. Consider this example:
```rust
// Safety invariant:
// - when SIGNAL_FLAG is false, only the signal handler can access SIGINFO.
// - when SIGNAL_FLAG is true, only the main thread can access SIGINFO.
#[thread_local]
static mut SIGINFO: siginfo_t = unsafe { mem::zeroed() };
#[thread_local]
static SIGNAL_FLAG: AtomicBool = AtomicBool::new(false);

extern "C" fn signal_handler(sig: i32, siginfo: *mut siginfo_t, ucontext: *mut c_void) {
    // This is atomic if the signal handler is set up so all signals are blocked
    // while the signal handler is running.
    if SIGNAL_FLAG.load(Relaxed) == false {
        unsafe {
            SIGINFO = *siginfo;
        }
        SIGNAL_FLAG.store(true, Relaxed);
    } else {
        // If there is already a signal stashed, ignore the new one.
    }
}

fn check_for_signal() -> Option<siginfo_t> {
    if SIGNAL_FLAG.load(Relaxed) {
        compiler_fence(Acquire);
        let siginfo = unsafe { SIGINFO };
        compiler_fence(Release);
        SIGNAL_FLAG.store(false);
        Some(siginfo)
    } else {
        is_none
    }
}
```

Note that `check_for_signal` could have been written using `fence` instead of `compiler_fence`, or even with `load(Acquire)` and `store(Release)` and it would still be correct: this is because `compiler_fence` is really just a subset of `fence` with a more limited synchronization scope.

The question of volatile accesses is interesting and I think the best way to make them fit into this model is to consider a volatile write to consist of one or more relaxed atomic stores. Similarly a volatile read consists of one or more relaxed atomic loads. Of course volatile accesses have additional semantics on top of that, but this suffices for the purposes of describing their interaction with fences.

**Comment by Lokathor** [comment]
For what it's worth LLVM puts volatile at the very start of the if-else evaluation of what a read of a particular byte returns when you read it. Volatile kicks in before all other consideration. The exact results are "target defined", and while volatile is specifically intended to support `sig_atomic_t`, it also "does not generally provide cross-thread synchronization".

Which all seems unfortunately vague to me.

[link](https://llvm.org/docs/LangRef.html#memory-model-for-concurrent-operations)

**Comment by RalfJung** [comment]
So in terms of the fence, this would all be resolved if we made the volatile operations also relaxed-atomic? That is something I have been advocating for a long time so this reaffirms my position.

With bytewise atomic memcpy coming in C++, can we declare them bytewise atomic relaxed operations? And does LLVM have a way of expressing that?

**Comment by Lokathor** [comment]
I don't know about the atomic parts, but you certainly can't split a native width access down into the individual byte accesses.

**Comment by RalfJung** [comment]
Oh yeah for volatile purposes it would still tear (or not) just like today. We don't specify it, which bothers me, but nobody is complaining too loudly so I guess it's okay.

I am just saying for the purpose of the atomic memory model, this would be bytewise atomic.

**Comment by Lokathor** [comment]
as long as the native access widths don't tear when aligned, that's enough. LLVM says it does this, we could make rust also say at least that much at some point.

**Comment by chorman0773** [comment]
IMO, an signal fence makes a ton of sense in a model with signal handlers. It's possible to execute code asynchronously (in the compilers logic) w/o leaving the thread of execution. A weaker form of synchronization that only allows communication with such signal handlers is perfectly valid.

Under that model, `compiler_fence` would make more sense if rust did acknowledge the fact that `compiler_fence` is `signal_fence`, not compiler-reordering barrier. Saying that `compiler_fence` is a reordering barrier implies that it is stronger than a `fence`, yet still weaker. However, fitting it into the memory model, it makes perfect sense.

**Comment by Lokathor** [comment]
Shouldn't we make `signal_fence` a separate thing, that clearly has those semantics the whole time, so that code intending for those specific (very limited) benefits can opt into that? That is, we've got a lot of older code which might be expecting `compiler_fence` to do _more_ than what you just described, and so weakening the existing `compiler_fence` might cause issues.

In terms of making sure that older code gets updated to the right semantics, should we make a new `signal_fence` function and then possibly deprecate `compiler_fence` if it can't do what people currently think it does?

**Comment by chorman0773** [comment]
I don't think that `compiler_fence` can do what you describe, no, but at the same time, I'm not sure what it would need to do.

Wouldn't most models with synchronous hardware interrupts treat them as asynchronous signals (this is what SNES-Dev does)?

**Comment by Lokathor** [comment]
Amanieu has in the past advised me to "treat the interrupt handler as a separate thread", so I've always tried to think about it in that way. That's actually probably *too* strict compared to what's needed on bare-metal single-core devices that don't have out-of-order execution (SNES, GBA, etc), but even for them it does basically work as a mental model.

**Comment by Amanieu** [comment]
I'll repeat my position on `compiler_fence` from [this previous comment](https://github.com/rust-lang/unsafe-code-guidelines/issues/347#issuecomment-1183591879):

> Anyways, getting back on topic, the way I view `compiler_fence` is that it works exactly like `fence` except that it only synchronizes with other threads on the same CPU. In practice "same CPU" include signal handlers or interrupt handlers which act as another thread from the AM's point of view (which also suspends the original thread).

**Comment by chorman0773** [comment]
My question is how we define "same CPU" beyond signal and interrupt handlers. Same physical/logical CPU? I think it may not be contrived to imagine an architecture where the kernel saving context can also separate synchronization effects at a hardware level, such that only the signal handler will see `atomic_signal_fence` synchronization. 
This is almost possible, I believe, on my own architecture, Clever-ISA, with a few additional extensions, but for the fact that all memory reads are already `thread_fence`-level acquire and memory writes are `thread_fence`-level release.

Incidentally, signal handlers most definitely are not guaranteed to execute on the same logical CPU as a thread of execution. The kernel can 100% schedule the execution of a signal handler on a different CPU, and, by luck of the draw, schedule only the signal handler.

**Comment by bjorn3** [comment]
> This is almost possible, I believe, on my own architecture, Clever-ISA, with a few additional extensions, but for the fact that all memory reads are already thread_fence-level acquire and memory writes are thread_fence-level release.

On pretty much every architecture loads and stores are thread_fence-level seqcst, right? Do you mean that this is not the case on your ISA? If that is the case wouldn't the OS need to insert a fence?

> Incidentally, signal handlers most definitely are not guaranteed to execute on the same logical CPU as a thread of execution. The kernel can 100% schedule the execution of a signal handler on a different CPU, and, by luck of the draw, schedule only the signal handler.

That would still imply a full synchronization at the kernel level to ensure that the thread receiving the signal has stopped before the signal handler runs.

**Comment by chorman0773** [comment]
> On pretty much every architecture loads and stores are thread_fence-level seqcst, right? Do you mean that this is not the case on your ISA? If that is the case wouldn't the OS need to insert a fence?

I know no architecture where every single load and store combined forms a global total order. Relaxed on the memory location? Yes. But x86 isn't even that strong as to SeqCst everything (though at least some x86s may be store-SeqCst, and `lock rmw`s are SeqCst).  `thread_fence` refers to the actual `fence` call, or the C++ `atomic_thread_fence`.

Some extensions of Clever that have been planned (though have not yet been proposed in any official capacity) might allow memory reads and writes to be deferred, and then committed after a context switch. Use of these instructions in codegen may be beneficial, but require the same deferred-operation register groups in order to reliably observe the operation comitting. The design has yet to begun (it won't in any significant capacity until the extension is proposed), but I can imagine that being tied to some register that gets saved-and-restored during a context switch (possibly even the virtual address resolution register), so that different threads can use deferred memory operations w/o interfering.

**Comment by bjorn3** [comment]
> I know no architecture where every single load and store combined forms a global total order. Relaxed on the memory location? Yes. But x86 isn't even that strong as to SeqCst everything (though at least some x86s may be store-SeqCst, and lock rmws are SeqCst). thread_fence refers to the actual fence call, or the C++ atomic_thread_fence.

Right, got confused with atomic_signal_fence.

> Some extensions of Clever that have been planned (though have not yet been proposed in any official capacity) might allow memory reads and writes to be deferred, and then committed after a context switch.

I see. In that case I think it would make sense to make the OS put a fence right before and after a signal handler runs. This allows keeping an atomic_signal_fence to be a no-op at the isa level and ensures that code which used volatile operations for communication with a signal handler when atomic_signal_fence didn't exist yet continues to work.

**Comment by chorman0773** [comment]
> I see. In that case I think it would make sense to make the OS put a fence right before and after a signal handler runs. This allows keeping an atomic_signal_fence to be a no-op at the isa level and ensures that code which used volatile operations for communication with a signal handler when atomic_signal_fence didn't exist yet continues to work.

Specifically, the "fence" would be restoring the deferred operation context. 

That being said, we still have yet to define "same CPU" in a meaningful way to all of the examples I mentioned.

**Comment by mkroening** [comment]
Just as a data point, I would like to mention that I just published [`InterruptRefCell`](https://docs.rs/interrupt-ref-cell), which uses a compiler fence to synchronize with signal handlers and interrupt handlers to share a `RefCell` with them.

**Comment by Lokathor** [comment]
If compiler fences are currently of dubious status I'm not sure I'd rely on that crate too carefully.

**Comment by mkroening** [comment]
> If compiler fences are currently of dubious status I'm not sure I'd rely on that crate too carefully.

I don't think, we can do any better currently and I needed for such an abstraction _now_. The crate wraps `RefCell` and does not add any unsafe code. Synchronization and mutual exclusion with signal handlers are quality of life features.

If you have any concrete issues with the crate, I would be happy about you opening an issue. :)
I am not sure how relevant this discussion is here. If you feel like it is, please continue, though.

**Comment by RalfJung** [comment]
Signal handlers (and interrupts which are basically the same thing) are the one place where I can actually make sense of compiler fences. We just need to adopt the C++ semantics that signal handlers are "weird threads", and compiler fences are fences that can be used to synchronize with them. (C++ calls them `atomic_signal_fence` which represents their purpose a lot better than Rust's name.)

However one still needs to use atomic memory operations to get any synchronization; non-atomic operations cause a data race, so your RefCell might still be UB @mkroening.

**Comment by DemiMarie** [comment]
Why do non-atomic operations cause a data race?  C explicitly states that using `volatile sig_atomic_t` to communicate with a signal handler is well-defined, and `volatile` is equivalent to a compiler fence (on every access, and with respect to other `volatile` variables only).  I agree that one needs to use either atomic _or_ volatile operations so that the compiler knows that the interrupt handler might access the data, but if the interrupt handler will always interrupt the same thread that the data is being accessed from, volatile should be sufficient.

**Comment by chorman0773** [comment]
It's specifically `volatile sig_atomic_t` that is exempt, beyond lock-free atomics.

C++17 signal handling rules
> Expressions executed on a signal handler are unsequenced with other expressions executed on the same thread of execution (except those executed as part of the same signal handler invocation), except that accesses through glvalues of type `volatile sig_atomic_t` are indeterminately sequenced with all other accesses through such glvalues on the same thread of execution. The functions `raise` and `abort` *synchronize-with* the beginning and end of the invocation of the signal handler caused by those functions.

**Comment by RalfJung** [comment]
General `volatile` still causes data race since sufficiently large volatile accesses can tear and the signal handler might be run in the middle of the access.

My preference would be to have some sort of "atomic volatile" primitive (we had proposals for that in this issue tracker), which -- by virtue of being atomic -- then automatically integrates with the concurrency system. The existing volatile accesses are then specified to emit a sequence of atomic volatile accesses, where e.g. all accesses no larger than a pointer are guaranteed to be just a single access but beyond that we make no promises about how the large access is split into one or more atomic volatile accesses.

**Comment by Lokathor** [comment]
It's actually slightly more finnicky than that because some types are smaller than a pointer but still can't be done in a single access (eg: `[u8; 3]` on ARM is 3 bytes, a pointer is 4 bytes, but the whole array can't in general be accessed in one access).

But that's a good general direction to move in, yes.

**Comment by RalfJung** [comment]
Ah right, I should have said "accesses with a power-of-two size that are no larger than a pointer".

**Comment by chorman0773** [comment]
That's also still not the case everywhere.

For example, on w65, pointers are 4 bytes, but the maximum size that can be written/read in one operation is 2 bytes, so pointer ops can still definitely tear.

**Comment by Lokathor** [comment]
I think "can be performed without tearing", and having that be determined by whatever CPU, is maybe the best we can do.

And sometimes it doesn't even matter if things might tear. Like the programmer may have additional knowledge that allows them to not worry about tearing "right here" because they know it's impossible in that situation (interrupts disabled on a single-core device, etc etc).

**Comment by briansmith** [comment]
> Signal handlers (and interrupts which are basically the same thing) are the one place where I can actually make sense of compiler fences. 

Consider this almost-real-world-code:
```
use std::{
    fs, io,
    os::fd::{IntoRawFd as _, RawFd},
    sync::{
        atomic::{compiler_fence, AtomicI32, Ordering},
        Mutex,
    },
};

fn get_lazy_fd() -> Result<RawFd, io::Error> {
    const FD_UNINIT: i32 = RawFd::MIN;

    // FD isn't stored within the mutex so we can usually use it
    // without
    static FD: AtomicI32 = AtomicI32::new(FD_UNINIT);
    static LOCK: Mutex<()> = Mutex::new(());
    match FD.load(Ordering::Acquire) {
        fd if fd != FD_UNINIT => Ok(fd),
        _ => {
            let _guard = LOCK.lock().unwrap();
            let fd = fs::File::open("/dev/urandom")?.into_raw_fd();
            FD.store(fd, Ordering::Release);
            
            // Force `FD` to be writen before we release the lock.
            // `Mutex` doesn't guarantee it will wait for the above store
            // before unlocking, because `FD` isn't within the Mutex.
            compiler_fence(Ordering::Release);

            Ok(fd)
        }
    }
}
```
Note in particular that `Mutex` doesn't specify its synchronization properties at all, and in particular it doesn't specify how it synchronizes with reads/writes to atomics outside of its contents. So potentially a compiler fence could help here.

**Comment by bjorn3** [comment]
> `Mutex` doesn't guarantee it will wait for the above store before unlocking, because `FD` isn't within the Mutex.

There is no such thing as being inside or outside the Mutex. If you have a Box, the target of this Box is not inside the Mutex value, yet has to have the synchronization of Mutex apply to it for soundness. A Mutex in rust is semantically a wrapper around a pthread_mutex_t (or whatever the platform equivalent is) and a value. While it ensures that you can't access the value while you haven't locked it, there is nothing tying the memory ordering of the inner mutex with just the wrapped value rather than everything the current thread reads/writes.

**Comment by briansmith** [comment]
@bjorn3 That's not what's documented, anywhere, so I don't rely on it.

**Comment by chorman0773** [comment]
I'd almost expect the compiler_fence *not* to do anything here. If the Mutex only locally synchronizes the value (say, uses something akin to `consume`), then I wouldn't expect adding a compiler acquire/release barrier to do anything to synchronize between threads still.

**Comment by RalfJung** [comment]
yeah, if the Mutex doesn't provide the desired guarantee there's no way `compiler_fence` would help.

**Comment by briansmith** [comment]
> A compiler fence is also not documented to do anything like this, so you can't rely on that for sure.

The documentation says "With [Release](https://doc.rust-lang.org/std/sync/atomic/enum.Ordering.html#variant.Release), preceding reads and writes cannot be moved past subsequent writes."

So, if we take the small leap that unlocking the mutex requires some kind of write, at least, then the compiler fence will prevent the mutex from being unlocked, no?

> If you think more documentation is needed for a guarantee like that, that should be added to `Mutex`.

I would love to. But, the way I was planning to specify it is as follows:
* `Mutex::lock()` does `compiler_fence(Ordering::Acquire)` or stronger synchronization.
* Dropping the `MutexGuard` does `compiler_fence(Ordering::Release)` or stronger synchronization.

Of course, we could also specify that it does some kind of atomic load with `Acquire` (or stronger) semantics during locking and some kind of atomic store with `Release` (or stronger) semantics, but that would be both overly-vague and also over-specifying the behavior, IMO.

The only other possibility is that `Mutex` (and `OnceLock`) have (C++20) consume semantics, instead of acquire/release, which is actually quite reasonable, but probably isn't compatible with existing usage.

**Comment by chorman0773** [comment]
Two issues:
1. The reordering model is not normative, atomic orderings affect visible side effects.
2. compiler_fence performs no actions that synchronize operations on any thread with another thread. In particular, compiler_fence typically generates no code, include code required to prevent the processor from interfering in ordering across hardware threads.

**Comment by briansmith** [comment]
> Two issues:
> 
> 1. The reordering model is not normative, atomic orderings affect visible side effects.
> 2. compiler_fence performs no actions that synchronize operations on any thread with another thread. In particular, compiler_fence typically generates no code, include code required to prevent the processor from interfering in visibility across hardware threads.

I see what you're saying. Importantly, for any clarification of `Mutex ` Acquire/Release semantics, we can do it in terms of `atomic::fence` instead of `compiler_fence`.

So it would instead need to be more like:
* `Mutex::lock(`) does `std::sync::atomic::fence(Ordering::Acquire)` or stronger synchronization.
* Dropping the `MutexGuard` does `std::sync::atomic::fence(Ordering::Release)` or stronger synchronization.

...which makes it out of scope for this issue.

OTOH, I think the way `compiler_fence` uses the words "read" and "write" is confusing because I consider any atomic loads/stores to be "read"/"write" operations, respectively. So I think `compiler_fence` must do the same thing as `fence` w.r.t. ordering all atomic load/stores, but not necessarily w.r.t. non-atomic reads/writes.

**Comment by chorman0773** [comment]
The definition of `compiler_fence` is mostly nonsencial IMO. The closest definition I have is the one I have above, which is that it is equivalent to `atomic_signal_fence` from C++ - IE. it will perform the same synchronizations as `fence`, but will only attach to atomic operations and other fences on the same thread of execution (possibly being executed by a signal handler).

**Comment by briansmith** [comment]
>  will perform the same synchronizations as fence, but will only attach to atomic operations and other fences (including compiler_fences) on the same thread of execution (possibly being executed by a signal handler).

This could be simplified by saying that:
* A compiler fence only attaches to other compiler fences, only on the same thread, 
* An `atomic::fence` is a `compiler_fence` / interacts with compiler fences as though it is also a compiler fence (define the mapping for each `Ordering` value.
* atomic load with `Ordering::Acquire`, `Mutex::[try_]lock()`, `Once::call_once_*` (before calling the init function), `OnceLock::get_*` (before calling the init function), etc. all act like `atomic::fence()` (with the same mapping to `compiler_fence`), and analogously for `Ordering::Release` with the corresponding operations, etc.
* Would `OnceCell` would have any interaction with `compiler_fence` analogously to how a `OnceLock` has to interact with atomic fences? It kind of seems like it would make sense by analogy above. Either way, it's worth explicitly documenting the (lack of) interaction.

Then every time we define a new thing that acts like an atomic fence, we would get the documentation for the interaction with compiler fences "for free."

Note that this could be done regardless o whether compiler fence is effectively renamed to something that indicates it is signal-handler specific.

**Comment by chorman0773** [comment]
fence including compiler_fence is fine.

All non-Relaxed atomics being fences does not work, since a fence has affects that spply to multiple atomic accesses, whereas an aromic access alone only forms synchronizes-with between actual access/load pairs.

**Comment by chorman0773** [comment]
> A release fence A synchronizes with an acquire fence B if there exist atomic operations X and Y, both operating on some atomic object M, such that A is sequenced before X, X modifies M, Y is sequenced before B, and Y reads the value written by X or a value written by any side effect in the hypothetical release sequence X would head if it were a release operation[.](https://eel.is/c++draft/atomics.fences#2.sentence-1)

(and similar provisions for synchronizing with atomic operations on both sides)

Notably, there's no restriction on which `X`s are chosen, so it applies to allow possibly valid `X` and `Y` pairs.

**Comment by RalfJung** [comment]
>  Importantly, for any clarification of Mutex Acquire/Release semantics, we can do it in terms of atomic::fence instead of compiler_fence.

Yes... though we have to be a bit careful in the wording here. Mutex acquire can be implemented entirely with acquire loads, not using any acquire fence, so I don't think a fence-based explanation can work. Instead the clarification needs to be in terms of which happens-before edges are guaranteed to exist.

But as you said that's getting off-topic for this issue.

> The documented definition of compiler_fence is mostly nonsencial IMO.

Agreed. Explanations in terms of reordering may be good for getting some first intuition, but they are unsuited as normative definitions. And the only normative definition for compiler_fence that I know of is C++'s `atomic_signal_fence`.

**Comment by briansmith** [comment]
Regarding clarifying how a Mutex works, I filed https://github.com/rust-lang/rust/issues/126239.

**Comment by RalfJung** [comment]
Seen on [LWN](https://lwn.net/Articles/985910/):

> My understanding is that compiler fences can be used instead of thread fences if you know that two threads only ever run on the same physical CPUs. So, as far as the compiler is concerned, they should block the same optimizations as thread fences, generating the same code apart from the fence instructions themselves. The assumptions that are made on generated instructions might be dubious in terms of portability, but they're fine with respect to data races and hence UB.

That's not how compiler fences are specified, but I wonder -- does anything go wrong with this approach?

**Comment by Diggsey** [comment]
That sounds right - the "same physical CPU" doesn't have meaning inside the AM AIUI, but I think we all expected that there would be a non-AM component to the definition of a compiler fence.

I guess a follow-up question is "what counts as the same physical CPU?". Is it a logical CPU core, a physical CPU core, a physical CPU, or just left up to the platform to decide (in which case it would be good to have answers for x86[-64] and ARM64 at least.)

**Comment by RalfJung** [comment]
> I guess a follow-up question is "what counts as the same physical CPU?".

It's something that behaves entirely in-order.

So when you run in userspace, a single thread (with its signal handler) is "the same physical CPU". I would say that if you spawn a thread with `thread::spawn`, that is *not* the same "physical CPU" *even if* you pin all threads to one core; this is because the AM sees you create that thread so it may assume it to be truly an independent thread.

On hardware, it's whatever the unit of "sequential computation" in this hardware is. If "logical CPU core" refers to things like hyperthreading, then I think that's what it is.

**Comment by Diggsey** [comment]
> I would say that if you spawn a thread with thread::spawn, that is not the same "physical CPU" even if you pin all threads to one core; this is because the AM sees you create that thread so it may assume it to be truly an independent thread.

That doesn't seem like it would enable any optimizations, since the compiler is not necessarily aware of what signal handlers exist, and so would still have to restrict re-orderings on both threads (even assuming they execute on different cores) under the presumption that both threads have a signal handler.

Given that, it seems as though a user could effectively rely on compiler fences working cross-thread (on the same logical CPU) even if we don't guarantee that explicitly.

> On hardware, it's whatever the unit of "sequential computation" in this hardware is. If "logical CPU core" refers to things like hyperthreading, then I think that's what it is.

Yeah, I guess the pathological case for hyper-threading is that despite two logical cores sharing the same L1 cache, the cores execute instructions out-of-order, breaking any guarantees of the fence.

**Comment by hanna-kruppe** [comment]
I have a use case for compiler fences that (as far as I can tell) was not yet discussed in this issue, and where I'd want compiler fences to mean *something* even across CPU cores: stack sampling for profiling, especially in bytecode interpreters.

The gold standard way of building sampling profilers is to periodically interrupt a currently running thread, and take a snapshot of its PC, call stack, etc. before you let it continue. Doing it this way requires [some fiddly and platform-specific code to suspend threads](https://nikhilism.com/post/2018/sampling-profiler-internals-suspending-threads/) but avoids large, systematic inaccuracies [introduced by only sampling at certain "safe points" in the program](https://plv.colorado.edu/papers/mytkowicz-pldi10.pdf). For this to work well, you need two kinds of "synchronization" in a muddy, conventional sense:

1. Suspending the target threads needs to "happen before" (not necessarily in the AM sense) the act of taking a snapshot. In other words, you actually suspend the thread before you take a snapshot, so you don't have to worry about data races while taking the snapshot.
2. The snapshot should, as much as possible, give a coherent view of some program point at the source level. While we want the program to run normally and enjoy any optimizations usually applied to it, there are a few "reorderings" that would be very problematic for sampling.

The first point is generally ensured by the platform-specific facilities used to suspend a thread. Only the second point is related to compiler fences. For example, setting up a stack frame during a call is not "atomic" in any real sense, but it's better if the stack pointer is (from profiler's POV) only adjusted *after* the return address is written to the expected position relative to the new stack pointer. An AOT or JIT compiler that generates machine code directly, or an interpreter written in assembly, can achieve this by just being careful about the sequence of CPU instructions it uses to manipulate the call stack. Because the thread will be completely interrupted when the sample is taken, this can use conventional non-atomic loads/stores even on machines with very weak memory models (as long as they have precise interrupts). After all, that's how any sampling profiling of C, C++, Rust, etc. code works.

But a bytecode interpreter written in Rust, which wants to permit cross-thread sampling of the bytecode call stack, needs to be more careful. There's basically three options:

1. Use inline assembly for those bits of the interpreter, which could otherwise be pure Rust (and in many cases, 100% safe Rust). This would force some tricky platform-specific bits into a core part of the interpreter, and the necessary memory clobbers may inhibit a lot of compiler optimizations that would be unproblematic and desirable.
2. Come up with some clever scheme to guarantee all the synchronization you want using only regular atomics and fences. This is likely to cause overhead: call stack manipulation happens frequently, and the bytecode PC (which is also sampled) is updated even more frequently. There's no contention but even uncontended atomics [are measurably more expensive than the non-atomic equivalent](https://travisdowns.github.io/blog/2020/07/06/concurrency-costs.html) if you do enough of them.
3. Use only relaxed loads and stores (no RMW) plus compiler fences, hoping that those will be cheaper than the previous option while still effectively guaranteeing the desired behavior on the platforms where you've also implemented the "interrupt a thread" bit.

The third option might look like this (on the interpreter side; assume the code that takes the snapshot has a matching acquire fence):

```rust
struct Interpreter {
    // all Arcs shared with the profiler thread
    pc: Arc<AtomicU32>,
    sp: Arc<AtomicU32>,
    stack: Arc<[AtomicU32]>
}

impl Interpreter {
    fn exec_call(&mut self, new_pc: u32) {
        let sp = self.sp.load(Relaxed);
        let old_pc = self.pc.load(Relaxed);
        self.stack[sp as usize].store(old_pc, Relaxed);
        self.pc.store(new_pc, Relaxed);
        // SP increment needs to happen after storing the new return address
        compiler_fence(Release);
        self.sp.store(sp + 1, Relaxed);
    }

    fn exec_return(&mut self) {
        let sp = self.sp.load(Relaxed);
        let new_pc = self.stack[(sp - 1) as usize].load(Relaxed);
        self.pc.store(new_pc, Relaxed);
        self.sp.store(sp - 1, Relaxed);
        // SP decrement needs to happen before any writes to the stack from a subsequent call.
        compiler_fence(Release);
    }
}
```

Note that this is all safe code, taking the snapshot also uses relaxed loads, so the problem is only possible race conditions. The argument why this sort of thing should work necessarily depends on the platform-specific bits now shown above. On Linux, suspending is generally done by delivering a signal to the thread you want to interrupt. The signal handler takes the snapshot and uses proper synchronization primitives (e.g., POSIX semaphores because those are async-signal-safe) to talk to the profiler thread. If a compiler fence is like a regular fence restricted to a signal handler and the thread it interrupts, you can probably argue for the correctness almost entirely in Rust memory model terms. Or disprove it, if I made a mistake in the example code.

But on Windows and macOS, the preferred/only(?) way is to ask the OS to suspend the target thread entirely and take the sample from *another* thread, which will generally run on another CPU core. Even outside of AM concerns, thinking only about the CPU architecture, it's not obvious that this can be correct. However, in practice it should be fine:

1. Once the OS has suspended the target thread by proactively interrupting it or waiting for its time slice to run out, the CPU has committed to the illusion of stopping at the boundary between two adjacent instructions, similar to how a signal handler might be invoked. If well-placed compiler fences restrict the translation from AM to machine code sufficiently for a signal handler to make sense of the program state in this situation, the only additional difficulty for taking a snapshot from another CPU core should be due to what that means for the CPU memory model.
2. Within the CPU memory model, it's very likely that the combination of everything the OS does establishes as much synchronization as needed. That includes interrupting the target thread (if it was running), doing the front half of a context switch (if the target thread was running), and somehow getting the confirmation that it's done to the profiler thread (and the core it's running on).

Because it's all safe code in the end (well, except the "suspending a running thread" bit, but that's a different matter) I am not exactly losing sleep over it. Still, this is a very gray area and I'd love to have more official ammunition for this way of using compiler fences. Alternatively, a convincing argument why compiler fences are the wrong tool for this and I need to do X instead would also be welcome.

**Comment by hanna-kruppe** [comment]
I've not seen any OS documentation that comes out and says anything like that. This is a very sparsely documented area, e.g., it took me a whole lot of searching to learn that Windows requires you to call `GetThreadContext` (even if you don't care about its contents) because [just calling `SuspendThread` won't wait for the thread to actually be suspended](https://stackoverflow.com/a/3515613). I still don't know if there's an equivalent gotcha on macOS, because its equivalent is even less documented.

I agree that migrating a thread between cores faces a similar problem and operating systems solve it somehow. In practical terms, it also seems very likely that they make "suspend another thread and read its thread state" apply enough synchronization (whatever that means on the CPU in question) to make basic tools like profilers work. However, it's not exactly the same as migrating a single thread and I'm not totally sure if that complicates anything under sufficiently weak memory models. For example, if thread migration requires some strong barrier instructions both on the old and the new core, then the OS would also have to do that on the other core when reporting that the thread was suspended, even if a weaker barrier was sufficient for basic message passing. Or maybe there's a fast path that checks a "is this thread running?" flag without necessarily creating enough synchronization. In this case, the semantics of compiler fence might affect whether it needs to emit machine code or not on the relevant platforms.

**Comment by RalfJung** [comment]
A compiler fence never has to emit machine code. That's pretty much its defining property. ;)

Fundamentally, the OS needs to provide *some* operation that is guaranteed to sync with a previously suspended thread (or make the "suspend a thread" function do that immediately). And it needs to be strong enough to do everything a migration of that thread to the current core would do. I understand OSes don't document that, but this is not something Rust is much involved with. Once that is given, one can make an argument that the "inspecting" thread has basically "taken over the logical core" of the previously running thread and therefore a compiler fence is sufficient.

**Comment by hanna-kruppe** [comment]
Well, in practice it sometimes does emit machine code (https://github.com/rust-lang/rust/issues/62256) and a sufficiently cursed target machine may need some extra code just to implement whatever Rust ends up choosing as the meaning of "compiler fences".

In any case, I see what you're saying. If compiler fences were defined with respect to logical CPU cores, one could make that arguments. I wouldn't love relying on it, but it would be better than if compiler fences only meant anything within a single AM thread (because I don't see how one could argue about *that*).

**Comment by RalfJung** [comment]
> Well, in practice it sometimes does sometimes emit machine code (https://github.com/rust-lang/rust/issues/62256)

That seems to bottom out [here](https://github.com/rust-embedded/wg/issues/361#issue-457900783). I don't know what `sw` does, but no hardware fence should be needed here. Ofc the fence can change which instructions are emitted for surrounding operations.

But anyway that's kind of off-topic here.

> In any case, I see what you're saying. If compiler fences were defined with respect to logical CPU cores, one could make that arguments. I wouldn't love relying on it, but it would be better than if compiler fences only meant anything within a single AM thread (because I don't see how one could argue about that).

The challenge for the docs is how to define "logical CPU core". But other than that, I think it is the right definition. So if someone wants to update the docs with an attempt at such a definition, please go ahead. :)

**Comment by hanna-kruppe** [comment]
For what it's worth, another use case for cross-thread compiler fences is when they're paired with another thread doing [membarrier](https://www.man7.org/linux/man-pages/man2/membarrier.2.html) syscalls (mentioned briefly earlier in the thread) or [other ways of achieving the same effect](https://github.com/pkhuong/barrierd). In those cases, you can't very well argue that the side calling membarrier has taken over the CPU core from the thread using compiler fences, because that thread is not suspended and may in fact be running concurrently. Clearly the OS + hardware establishes some sort of synchronization, but how does that connect to the compiler fence? Perhaps you could pretend that the OS scheduler interrupts issue both a compiler fence and a regular fence every time it fires on a logical core? That seems to imply much more synchronization happening than the AM would like to admit. Maybe it won't prevent any optimizations if userspace code can't rely on those fences because it can't know when they happen (for lack of any shared state with the interrupt).

**Comment by RalfJung** [comment]
Is that the one that fires a cross core interrupt to all other cores, i.e. some sort of asymmetric fence? I think that is a whole new primitive that needs to be added to the memory model as such.

**Comment by hanna-kruppe** [comment]
It is an asymmetric fence in a way, but IPI is only one possible way of implementing it. Linux membarrier has a way that piggy-backs on the RCU grace period (whatever that means, don't ask me). barrierd is a userspace + eBPF solution mostly based on waiting until any interrupt has fired on every core at least once (relying on those being serializing for the CPU).

**Comment by chorman0773** [comment]
Interrupts are a different conccurency context from the thread they're running on. Synchronizing between an interrupt handler running on a thread and another thread won't do anything to synchronize between the two threads.

**Comment by hanna-kruppe** [comment]
It's not obvious to me why there should be any difference between asking the OS to send my thread a signal every 10ms and doing something funky in the signal handler, or (in bare-metal code) setting up a timer interrupt for every 10 ms and doing the same funky stuff in the interrupt handler. It's not literally exactly the same, but both seem equally foreign to the AM and both have pretty much the same implications for the hardware memory model.

**Comment by DemiMarie** [comment]
There are accelerators on which one isn’t even guaranteed to read what one just wrote to memory unless one uses a barrier instruction.

**Comment by bjorn3** [comment]
> There are accelerators on which one isn’t even guaranteed to read what one just wrote to memory unless one uses a barrier instruction.

For such an accelerator to run Rust the compiler would need to insert a barrier between every read and every write. The C++ memory model enforces happens before within a single thread. If the cpu core doesn't guarantee this the compiler is required to insert barriers as necessary to enforce this.

**Comment by DemiMarie** [comment]
> > There are accelerators on which one isn’t even guaranteed to read what one just wrote to memory unless one uses a barrier instruction.
> 
> For such an accelerator to run Rust the compiler would need to insert a barrier between every read and every write. The C++ memory model enforces happens before within a single thread. If the cpu core doesn't guarantee this the compiler is required to insert barriers as necessary to enforce this.

Given that this would be ludicrously expensive, I suspect that the compiler would instead only allow local memory (which is usually coherent with program execution) to be accessed using Rust pointer syntax, and require access to incoherent memory to use function calls that are opaque to the compiler.

**Comment by DemiMarie** [comment]
This is off-topic, though.

**Comment by CAD97** [comment]
Based on the stack sampling example, I think it could be potentially reasonable to document `compiler_fence` as synchronization with the concept of thread suspension. Expositionally, if you can determine that a thread of execution is suspended (i.e. not currently executing) through some target-specific manner (e.g. using thread suspension APIs or being an interrupt running on that thread), then the observable state will be consistent with

- either no writes after any given `compiler_fence(Release)` have happened or all writes before that fence have, and
- either no reads after any given `compiler_fence(Acquire)` have happened or all reads after that fence have

but unfortunately it's on the thread suspension to ensure this is actually true on weak memory targets (e.g. by ensuring all weak operations commit before considering a thread to be fully suspended) as otherwise the weak operations could become visible out of order across the intended ordering of the `compiler_fence`.

TL;DR: that's a lot of words to say again that synchronizing within a logical CPU core must be transitive across synchronizing with thread suspension on any useful system.

**Comment by comex** [comment]
@arielb1 Which AcqRel fence?  Which load?

**Comment by RalfJung** [comment]
I think this is largely resolved -- our "compiler fences" are the same thing as signal fences in C++, we just picked an unfortunate name. The docs are fairly clear about that. There may be a need for other kinds of compiler fences, but those should get new issues with a clear problem statement.
